File: FreeLine-Android/app/src/main/java/com/freeline/app/calls/TwilioVoiceTransport.kt
package com.freeline.app.calls

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.twilio.audioswitch.AudioDevice
import com.twilio.audioswitch.AudioSwitch
import com.twilio.voice.Call
import com.twilio.voice.CallException
import com.twilio.voice.ConnectOptions
import com.twilio.voice.Voice

class TwilioVoiceTransport(
    context: Context,
) {
    private val appContext = context.applicationContext
    private val audioSwitch = AudioSwitch(appContext, null, true)
    private val mainHandler = Handler(Looper.getMainLooper())

    private var audioStarted = false
    private var currentCall: Call? = null
    private var eventHandler: ((VoiceCallEvent) -> Unit)? = null

    fun startOutgoingCall(
        token: String,
        to: String,
        eventHandler: (VoiceCallEvent) -> Unit,
    ) {
        check(isLikelyJwt(token)) {
            "Voice provider is not configured yet. Add Twilio voice credentials to the backend first."
        }

        this.eventHandler = eventHandler
        emit(VoiceCallEvent.Connecting)
        ensureAudioStarted()

        val connectOptions = ConnectOptions.Builder(token)
            .params(mapOf("to" to to))
            .build()

        currentCall = Voice.connect(appContext, connectOptions, object : Call.Listener {
            override fun onRinging(call: Call) {
                emit(VoiceCallEvent.Ringing)
            }

            override fun onConnectFailure(
                call: Call,
                callException: CallException,
            ) {
                currentCall = null
                audioSwitch.deactivate()
                emit(VoiceCallEvent.Failed(formatCallException(callException)))
            }

            override fun onConnected(call: Call) {
                currentCall = call
                audioSwitch.activate()
                setSpeakerEnabled(true)
                emit(VoiceCallEvent.Connected(System.currentTimeMillis()))
            }

            override fun onReconnecting(
                call: Call,
                callException: CallException,
            ) {
                emit(VoiceCallEvent.Reconnecting(formatCallException(callException)))
            }

            override fun onReconnected(call: Call) {
                emit(VoiceCallEvent.Reconnected)
            }

            override fun onDisconnected(
                call: Call,
                callException: CallException?,
            ) {
                currentCall = null
                audioSwitch.deactivate()
                emit(VoiceCallEvent.Disconnected(callException?.let(::formatCallException)))
            }

            override fun onCallQualityWarningsChanged(
                call: Call,
                currentWarnings: Set<Call.CallQualityWarning>,
                previousWarnings: Set<Call.CallQualityWarning>,
            ) {
                return
            }
        })
    }

    fun endActiveCall() {
        currentCall?.disconnect()
        currentCall = null
        audioSwitch.deactivate()
    }

    fun setMuted(isMuted: Boolean) {
        currentCall?.mute(isMuted)
    }

    fun setSpeakerEnabled(isEnabled: Boolean) {
        val nextDevice = if (isEnabled) {
            audioSwitch.availableAudioDevices.firstOrNull { it is AudioDevice.Speakerphone }
        } else {
            audioSwitch.availableAudioDevices.firstOrNull { it !is AudioDevice.Speakerphone }
        }

        if (nextDevice != null) {
            audioSwitch.selectDevice(nextDevice)
        }
    }

    fun sendDigits(digits: String) {
        if (digits.isNotBlank()) {
            currentCall?.sendDigits(digits)
        }
    }

    fun shutdown() {
        currentCall?.disconnect()
        currentCall = null
        audioSwitch.stop()
        audioStarted = false
    }

    private fun ensureAudioStarted() {
        if (!audioStarted) {
            audioSwitch.start()
            audioStarted = true
        }
    }

    private fun emit(event: VoiceCallEvent) {
        mainHandler.post {
            eventHandler?.invoke(event)
        }
    }

    private fun isLikelyJwt(token: String): Boolean = token.split('.').size == 3

    private fun formatCallException(callException: CallException): String {
        return "${callException.errorCode}: ${callException.message ?: "Call failed."}"
    }
}

File: FreeLine-iOS/Sources/Calls/TwilioVoiceTransport.swift
import AVFoundation
import Foundation
import TwilioVoice

enum VoiceTransportError: LocalizedError {
    case microphonePermissionDenied
    case providerNotConfigured

    var errorDescription: String? {
        switch self {
        case .microphonePermissionDenied:
            return "Microphone access is required before placing a call."
        case .providerNotConfigured:
            return "Voice provider is not configured yet. Add Twilio voice credentials to the backend first."
        }
    }
}

@MainActor
final class TwilioVoiceTransport: NSObject {
    private let audioDevice = DefaultAudioDevice()
    private var currentCall: Call?
    private var eventHandler: ((VoiceCallEvent) -> Void)?

    override init() {
        super.init()
        TwilioVoiceSDK.audioDevice = audioDevice
    }

    func startOutgoingCall(
        token: String,
        to remoteNumber: String,
        eventHandler: @escaping (VoiceCallEvent) -> Void
    ) async throws {
        guard token.split(separator: ".").count == 3 else {
            throw VoiceTransportError.providerNotConfigured
        }

        try await ensureMicrophonePermission()
        self.eventHandler = eventHandler
        eventHandler(.connecting)

        let connectOptions = ConnectOptions(accessToken: token) { builder in
            builder.params = ["to": remoteNumber]
        }

        currentCall = TwilioVoiceSDK.connect(options: connectOptions, delegate: self)
    }

    func endActiveCall() {
        currentCall?.disconnect()
        currentCall = nil
    }

    func setMuted(_ isMuted: Bool) {
        currentCall?.isMuted = isMuted
    }

    func setSpeakerEnabled(_ isEnabled: Bool) {
        audioDevice.block = {
            do {
                try AVAudioSession.sharedInstance().overrideOutputAudioPort(
                    isEnabled ? .speaker : .none
                )
            } catch {
                NSLog("Failed to toggle speaker route: \(error.localizedDescription)")
            }
        }

        audioDevice.block()
    }

    func sendDigits(_ digits: String) {
        guard !digits.isEmpty else {
            return
        }

        currentCall?.sendDigits(digits)
    }

    private func ensureMicrophonePermission() async throws {
        switch AVAudioApplication.shared.recordPermission {
        case .granted:
            return
        case .denied:
            throw VoiceTransportError.microphonePermissionDenied
        case .undetermined:
            let granted = await withCheckedContinuation { continuation in
                AVAudioApplication.requestRecordPermission { didGrant in
                    continuation.resume(returning: didGrant)
                }
            }

            guard granted else {
                throw VoiceTransportError.microphonePermissionDenied
            }
        @unknown default:
            throw VoiceTransportError.microphonePermissionDenied
        }
    }
}

extension TwilioVoiceTransport: @MainActor CallDelegate {
    func callDidStartRinging(call: Call) {
        eventHandler?(.ringing)
    }

    func callDidConnect(call: Call) {
        currentCall = call
        setSpeakerEnabled(true)
        eventHandler?(.connected(Date()))
    }

    func callIsReconnecting(call: Call, error: Error) {
        eventHandler?(.reconnecting(error.localizedDescription))
    }

    func callDidReconnect(call: Call) {
        eventHandler?(.reconnected)
    }

    func callDidFailToConnect(call: Call, error: Error) {
        currentCall = nil
        eventHandler?(.failed(error.localizedDescription))
    }

    func callDidDisconnect(call: Call, error: Error?) {
        currentCall = nil
        eventHandler?(.disconnected(error?.localizedDescription))
    }
}
